shared/src/main/scala/frankenpaxos/spaxosdecouple/Acceptor.scala

```diff
@@ -0,0 +1,384 @@
+      }
+    }
+  }
+
+  private def handleProposeRequest(
+      src: Transport#Address,
+      proposal: Proposal
+  ): Unit = {
+    metrics.requestsTotal.labels("ProposeRequest").inc()
+
+    // If the wait period and wait stagger are both 0, then we don't bother
+    // fiddling with a timer. We process the propose request immediately.
+    if (options.waitPeriod == java.time.Duration.ofSeconds(0) &&
+        options.waitStagger == java.time.Duration.ofSeconds(0)) {
+      processProposeRequest(src, proposal) match {
+        case Some(phase2b) =>
+          val fakeLeader = chan[FakeLeader[Transport]](src, FakeLeader.serializer)
+          fakeLeader.send(
+            FakeLeaderInbound().withPhase2BBuffer(Phase2bBuffer(Seq(phase2b)))
+          )
+
+        case None =>
+        // Do nothing.
+      }
+    } else {
+      val t = (System.nanoTime(), src, proposal)
+      bufferedProposeRequests += t
+    }
+  }
+
+  private def handlePhase1a(src: Transport#Address, phase1a: Phase1a): Unit = {
+    metrics.requestsTotal.labels("Phase1a").inc()
+
+    // Ignore messages from previous rounds.
+    if (phase1a.round <= round) {
+      logger.info(
+        s"An acceptor received a phase 1a message for round " +
+          s"${phase1a.round} but is in round $round."
+      )
+      val leader = chan[Leader[Transport]](src, Leader.serializer)
+      leader.send(
+        LeaderInbound().withPhase1BNack(
+          Phase1bNack(acceptorId = acceptorId, round = round)
+        )
+      )
+      return
+    }
+
+    // Bump our round and send the leader all of our votes. Note that we
+    // exclude votes below the chosen watermark and votes for slots that the
+    // leader knows are chosen. We also make sure not to return votes for slots
+    // that we haven't actually voted in.
+    round = phase1a.round
+    val votes = log
+      .prefix()
+      .iteratorFrom(phase1a.chosenWatermark)
+      .filter({ case (slot, _) => !phase1a.chosenSlot.contains(slot) })
+      .flatMap({
+        case (s, Entry(vr, VVCommand(command), _)) =>
+          Some(Phase1bVote(slot = s, voteRound = vr).withUniqueId(command))
+        case (s, Entry(vr, VVNoop, _)) =>
+          Some(Phase1bVote(slot = s, voteRound = vr).withNoop(Noop()))
+        case (_, Entry(_, VVNothing, _)) =>
+          None
+      })
+    val leader = leaders(config.roundSystem.leader(round))
+    leader.send(
+      LeaderInbound().withPhase1B(
+        Phase1b(acceptorId = acceptorId, round = round, vote = votes.to[Seq])
+      )
+    )
+  }
+
+  private def handlePhase2aBuffer(
+      src: Transport#Address,
+      phase2aBuffer: Phase2aBuffer
+  ): Unit = {
+    // TODO(mwhittaker): Add phase 2 nack.
+    metrics.requestsTotal.labels("Phase2aBuffer").inc()
+    val buffer = Phase2bBuffer(phase2aBuffer.phase2A.flatMap(processPhase2a))
+    val fakeLeader = chan[FakeLeader[Transport]](src, FakeLeader.serializer)
+    fakeLeader.send(FakeLeaderInbound().withPhase2BBuffer(buffer))
```

Comment:
If the acceptor nacks a particular instance, it should send the nack to the leader instead of the fake leader. We'll have to think about a good way to do that.